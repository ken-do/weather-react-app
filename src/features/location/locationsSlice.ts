import { createSlice, PayloadAction, createAction } from '@reduxjs/toolkit'
import { call, put, takeLatest } from 'redux-saga/effects'
import api, { endpoints } from 'utils/api'
import { Location } from '../../types/data'

interface LocationsState {
    isLoading: boolean
    entries: Location[]
    error: string | null
}

interface RequestResponse {
    data: Location[]
}

export const locationsInitialState: LocationsState = {
    isLoading: false,
    entries: [],
    error: null,
}

function reset(state: LocationsState) {
    state.isLoading = false
    state.entries = []
    state.error = null
}

function startLoading(state: LocationsState) {
    state.isLoading = true
}

function loadingFailed(
    state: LocationsState,
    { payload }: PayloadAction<Error>
) {
    state.isLoading = false
    state.error = payload.message
    state.entries = []
}

const locations = createSlice({
    name: 'locations',
    initialState: locationsInitialState,
    reducers: {
        resetLocations: reset,
        getLocationsStart: startLoading,
        getLocationsSuccess(state, { payload }: PayloadAction<Location[]>) {
            state.isLoading = false
            state.error = null
            state.entries = payload
        },
        getLocationsFailure: loadingFailed,
    },
})

export const getLocations = createAction<string>('locations/getLocations')

export const {
    resetLocations,
    getLocationsStart,
    getLocationsSuccess,
    getLocationsFailure,
} = locations.actions

export function* fetchLocations({ payload }: ReturnType<typeof getLocations>) {
    try {
        yield put(getLocationsStart())
        const response: RequestResponse = yield call(
            api.get,
            endpoints.locationSearch,
            {
                params: { query: payload },
            }
        )
        yield put(getLocationsSuccess(response.data))
    } catch (err) {
        yield put(getLocationsFailure(err))
    }
}

export function* watchFetchLocations() {
    yield takeLatest(getLocations.type, fetchLocations)
}

export default locations.reducer
