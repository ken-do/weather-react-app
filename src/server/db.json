{
    "search": [
        {
            "title": "San Francisco",
            "location_type": "City",
            "woeid": 2487956,
            "latt_long": "37.777119, -122.41964"
        },
        {
            "title": "San Diego",
            "location_type": "City",
            "woeid": 2487889,
            "latt_long": "32.715691,-117.161720"
        },
        {
            "title": "San Jose",
            "location_type": "City",
            "woeid": 2488042,
            "latt_long": "37.338581,-121.885567"
        },
        {
            "title": "San Antonio",
            "location_type": "City",
            "woeid": 2487796,
            "latt_long": "29.424580,-98.494614"
        },
        {
            "title": "Santa Cruz",
            "location_type": "City",
            "woeid": 2488853,
            "latt_long": "36.974018,-122.030952"
        },
        {
            "title": "Santiago",
            "location_type": "City",
            "woeid": 349859,
            "latt_long": "-33.463039,-70.647942"
        },
        {
            "title": "Santorini",
            "location_type": "City",
            "woeid": 56558361,
            "latt_long": "36.406651,25.456530"
        },
        {
            "title": "Santander",
            "location_type": "City",
            "woeid": 773964,
            "latt_long": "43.461498,-3.810010"
        },
        {
            "title": "Busan",
            "location_type": "City",
            "woeid": 1132447,
            "latt_long": "35.170429,128.999481"
        },
        {
            "title": "Santa Cruz de Tenerife",
            "location_type": "City",
            "woeid": 773692,
            "latt_long": "28.46163,-16.267059"
        },
        {
            "title": "Santa Fe",
            "location_type": "City",
            "woeid": 2488867,
            "latt_long": "35.666431,-105.972572"
        }
    ],
    "location": [
        {
            "consolidated_weather": [
                {
                    "id": 6018523798175744,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "N",
                    "created": "2021-02-07T06:20:19.776302Z",
                    "applicable_date": "2021-02-07",
                    "min_temp": 7.925,
                    "max_temp": 15.7,
                    "the_temp": 14.309999999999999,
                    "wind_speed": 2.7886128043134764,
                    "wind_direction": 355.28847761972753,
                    "air_pressure": 1017.0,
                    "humidity": 66,
                    "visibility": 13.206934005408415,
                    "predictability": 70
                },
                {
                    "id": 6136453466685440,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "NW",
                    "created": "2021-02-07T06:20:23.054980Z",
                    "applicable_date": "2021-02-08",
                    "min_temp": 8.57,
                    "max_temp": 14.04,
                    "the_temp": 13.649999999999999,
                    "wind_speed": 3.7480341844428544,
                    "wind_direction": 319.79960176305474,
                    "air_pressure": 1016.0,
                    "humidity": 66,
                    "visibility": 13.691914220949654,
                    "predictability": 70
                },
                {
                    "id": 5402338597535744,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "WNW",
                    "created": "2021-02-07T06:20:25.790143Z",
                    "applicable_date": "2021-02-09",
                    "min_temp": 9.004999999999999,
                    "max_temp": 12.184999999999999,
                    "the_temp": 13.655000000000001,
                    "wind_speed": 4.6433530153424,
                    "wind_direction": 281.56468032470184,
                    "air_pressure": 1016.5,
                    "humidity": 70,
                    "visibility": 13.566397240117713,
                    "predictability": 70
                },
                {
                    "id": 6728299423203328,
                    "weather_state_name": "Heavy Cloud",
                    "weather_state_abbr": "hc",
                    "wind_direction_compass": "W",
                    "created": "2021-02-07T06:20:30.641786Z",
                    "applicable_date": "2021-02-10",
                    "min_temp": 9.16,
                    "max_temp": 12.945,
                    "the_temp": 11.94,
                    "wind_speed": 5.112261200031814,
                    "wind_direction": 265.7516289079764,
                    "air_pressure": 1017.0,
                    "humidity": 73,
                    "visibility": 9.650826672234153,
                    "predictability": 71
                },
                {
                    "id": 4531863566680064,
                    "weather_state_name": "Heavy Cloud",
                    "weather_state_abbr": "hc",
                    "wind_direction_compass": "WSW",
                    "created": "2021-02-07T06:20:31.992459Z",
                    "applicable_date": "2021-02-11",
                    "min_temp": 8.2,
                    "max_temp": 12.094999999999999,
                    "the_temp": 12.61,
                    "wind_speed": 3.7801049868766405,
                    "wind_direction": 238.5,
                    "air_pressure": 1020.0,
                    "humidity": 70,
                    "visibility": 9.696497454863596,
                    "predictability": 71
                },
                {
                    "id": 4839661508755456,
                    "weather_state_name": "Heavy Rain",
                    "weather_state_abbr": "hr",
                    "wind_direction_compass": "WSW",
                    "created": "2021-02-07T06:20:34.700790Z",
                    "applicable_date": "2021-02-12",
                    "min_temp": 8.1,
                    "max_temp": 11.87,
                    "the_temp": 11.95,
                    "wind_speed": 6.434576871072934,
                    "wind_direction": 238.0,
                    "air_pressure": 1018.0,
                    "humidity": 68,
                    "visibility": 9.999726596675416,
                    "predictability": 77
                }
            ],
            "time": "2021-02-07T00:11:21.583156-08:00",
            "sun_rise": "2021-02-07T07:07:39.607285-08:00",
            "sun_set": "2021-02-07T17:39:20.933536-08:00",
            "timezone_name": "LMT",
            "parent": {
                "title": "California",
                "location_type": "Region / State / Province",
                "woeid": 2347563,
                "latt_long": "37.271881,-119.270233"
            },
            "sources": [
                {
                    "title": "BBC",
                    "slug": "bbc",
                    "url": "http://www.bbc.co.uk/weather/",
                    "crawl_rate": 360
                },
                {
                    "title": "Forecast.io",
                    "slug": "forecast-io",
                    "url": "http://forecast.io/",
                    "crawl_rate": 480
                },
                {
                    "title": "HAMweather",
                    "slug": "hamweather",
                    "url": "http://www.hamweather.com/",
                    "crawl_rate": 360
                },
                {
                    "title": "Met Office",
                    "slug": "met-office",
                    "url": "http://www.metoffice.gov.uk/",
                    "crawl_rate": 180
                },
                {
                    "title": "OpenWeatherMap",
                    "slug": "openweathermap",
                    "url": "http://openweathermap.org/",
                    "crawl_rate": 360
                },
                {
                    "title": "Weather Underground",
                    "slug": "wunderground",
                    "url": "https://www.wunderground.com/?apiref=fc30dc3cd224e19b",
                    "crawl_rate": 720
                },
                {
                    "title": "World Weather Online",
                    "slug": "world-weather-online",
                    "url": "http://www.worldweatheronline.com/",
                    "crawl_rate": 360
                }
            ],
            "title": "San Francisco",
            "location_type": "City",
            "woeid": 2487956,
            "latt_long": "37.777119, -122.41964",
            "timezone": "US/Pacific"
        },
        {
            "consolidated_weather": [
                {
                    "id": 5819995847131136,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "W",
                    "created": "2021-02-07T06:32:28.770578Z",
                    "applicable_date": "2021-02-07",
                    "min_temp": 11.309999999999999,
                    "max_temp": 18.884999999999998,
                    "the_temp": 18.295,
                    "wind_speed": 3.5211901904504366,
                    "wind_direction": 263.27478994815453,
                    "air_pressure": 1015.0,
                    "humidity": 56,
                    "visibility": 14.61340769903762,
                    "predictability": 70
                },
                {
                    "id": 4666365014179840,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "SW",
                    "created": "2021-02-07T06:32:31.894843Z",
                    "applicable_date": "2021-02-08",
                    "min_temp": 11.75,
                    "max_temp": 17.545,
                    "the_temp": 17.355,
                    "wind_speed": 4.005046050308863,
                    "wind_direction": 221.69782617594464,
                    "air_pressure": 1015.0,
                    "humidity": 61,
                    "visibility": 14.149243418436331,
                    "predictability": 70
                },
                {
                    "id": 6106026039312384,
                    "weather_state_name": "Heavy Cloud",
                    "weather_state_abbr": "hc",
                    "wind_direction_compass": "SW",
                    "created": "2021-02-07T06:32:34.786834Z",
                    "applicable_date": "2021-02-09",
                    "min_temp": 11.825,
                    "max_temp": 16.865000000000002,
                    "the_temp": 16.125,
                    "wind_speed": 3.671786735305814,
                    "wind_direction": 228.77822425191678,
                    "air_pressure": 1015.0,
                    "humidity": 66,
                    "visibility": 13.266585639863198,
                    "predictability": 71
                },
                {
                    "id": 5723463504363520,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "SW",
                    "created": "2021-02-07T06:32:38.363852Z",
                    "applicable_date": "2021-02-10",
                    "min_temp": 12.21,
                    "max_temp": 16.445,
                    "the_temp": 15.344999999999999,
                    "wind_speed": 4.226551565311912,
                    "wind_direction": 228.93331148161587,
                    "air_pressure": 1015.5,
                    "humidity": 69,
                    "visibility": 13.451132883957687,
                    "predictability": 70
                },
                {
                    "id": 6029439927320576,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "SW",
                    "created": "2021-02-07T06:32:41.472857Z",
                    "applicable_date": "2021-02-11",
                    "min_temp": 12.14,
                    "max_temp": 16.775,
                    "the_temp": 16.32,
                    "wind_speed": 3.8268653066094007,
                    "wind_direction": 214.5,
                    "air_pressure": 1015.0,
                    "humidity": 72,
                    "visibility": 9.999726596675416,
                    "predictability": 70
                },
                {
                    "id": 6081357626212352,
                    "weather_state_name": "Heavy Cloud",
                    "weather_state_abbr": "hc",
                    "wind_direction_compass": "WSW",
                    "created": "2021-02-07T06:32:44.063814Z",
                    "applicable_date": "2021-02-12",
                    "min_temp": 12.31,
                    "max_temp": 17.47,
                    "the_temp": 17.49,
                    "wind_speed": 6.8077362204724405,
                    "wind_direction": 251.5,
                    "air_pressure": 1015.0,
                    "humidity": 65,
                    "visibility": 9.999726596675416,
                    "predictability": 71
                }
            ],
            "time": "2021-02-07T00:14:24.651198-08:00",
            "sun_rise": "2021-02-07T06:38:34.873048-08:00",
            "sun_set": "2021-02-07T17:26:28.687579-08:00",
            "timezone_name": "LMT",
            "parent": {
                "title": "California",
                "location_type": "Region / State / Province",
                "woeid": 2347563,
                "latt_long": "37.271881,-119.270233"
            },
            "sources": [
                {
                    "title": "BBC",
                    "slug": "bbc",
                    "url": "http://www.bbc.co.uk/weather/",
                    "crawl_rate": 360
                },
                {
                    "title": "Forecast.io",
                    "slug": "forecast-io",
                    "url": "http://forecast.io/",
                    "crawl_rate": 480
                },
                {
                    "title": "HAMweather",
                    "slug": "hamweather",
                    "url": "http://www.hamweather.com/",
                    "crawl_rate": 360
                },
                {
                    "title": "Met Office",
                    "slug": "met-office",
                    "url": "http://www.metoffice.gov.uk/",
                    "crawl_rate": 180
                },
                {
                    "title": "OpenWeatherMap",
                    "slug": "openweathermap",
                    "url": "http://openweathermap.org/",
                    "crawl_rate": 360
                },
                {
                    "title": "World Weather Online",
                    "slug": "world-weather-online",
                    "url": "http://www.worldweatheronline.com/",
                    "crawl_rate": 360
                }
            ],
            "title": "San Diego",
            "location_type": "City",
            "woeid": 2487889,
            "latt_long": "32.715691,-117.161720",
            "timezone": "America/Los_Angeles"
        },
        {
            "consolidated_weather": [
                {
                    "id": 6049460112064512,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "NW",
                    "created": "2021-02-07T06:32:55.663480Z",
                    "applicable_date": "2021-02-07",
                    "min_temp": 7.455,
                    "max_temp": 19.145000000000003,
                    "the_temp": 16.895,
                    "wind_speed": 2.153696247782285,
                    "wind_direction": 322.9362319518955,
                    "air_pressure": 1016.5,
                    "humidity": 53,
                    "visibility": 13.601504712479121,
                    "predictability": 70
                },
                {
                    "id": 6052058516946944,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "N",
                    "created": "2021-02-07T06:32:58.968336Z",
                    "applicable_date": "2021-02-08",
                    "min_temp": 7.539999999999999,
                    "max_temp": 17.72,
                    "the_temp": 16.16,
                    "wind_speed": 2.3164520325326756,
                    "wind_direction": 2.3426995889576276,
                    "air_pressure": 1016.0,
                    "humidity": 57,
                    "visibility": 13.273731408573928,
                    "predictability": 70
                },
                {
                    "id": 5472824648007680,
                    "weather_state_name": "Light Cloud",
                    "weather_state_abbr": "lc",
                    "wind_direction_compass": "WNW",
                    "created": "2021-02-07T06:33:01.738210Z",
                    "applicable_date": "2021-02-09",
                    "min_temp": 8.58,
                    "max_temp": 15.8,
                    "the_temp": 15.73,
                    "wind_speed": 2.683768670487023,
                    "wind_direction": 286.69550186944974,
                    "air_pressure": 1016.0,
                    "humidity": 66,
                    "visibility": 13.469463334128688,
                    "predictability": 70
                },
                {
                    "id": 6579732964442112,
                    "weather_state_name": "Heavy Cloud",
                    "weather_state_abbr": "hc",
                    "wind_direction_compass": "WNW",
                    "created": "2021-02-07T06:33:04.696425Z",
                    "applicable_date": "2021-02-10",
                    "min_temp": 7.609999999999999,
                    "max_temp": 16.425,
                    "the_temp": 14.405,
                    "wind_speed": 3.1376863440547207,
                    "wind_direction": 286.66971818960064,
                    "air_pressure": 1017.5,
                    "humidity": 64,
                    "visibility": 10.260702497415096,
                    "predictability": 71
                },
                {
                    "id": 6598724554850304,
                    "weather_state_name": "Heavy Cloud",
                    "weather_state_abbr": "hc",
                    "wind_direction_compass": "WSW",
                    "created": "2021-02-07T06:33:08.358642Z",
                    "applicable_date": "2021-02-11",
                    "min_temp": 6.404999999999999,
                    "max_temp": 15.41,
                    "the_temp": 16.93,
                    "wind_speed": 2.3794780481985205,
                    "wind_direction": 248.5,
                    "air_pressure": 1020.0,
                    "humidity": 60,
                    "visibility": 9.843762427423844,
                    "predictability": 71
                },
                {
                    "id": 5543348950532096,
                    "weather_state_name": "Heavy Rain",
                    "weather_state_abbr": "hr",
                    "wind_direction_compass": "WSW",
                    "created": "2021-02-07T06:33:10.674810Z",
                    "applicable_date": "2021-02-12",
                    "min_temp": 7.300000000000001,
                    "max_temp": 14.295,
                    "the_temp": 14.7,
                    "wind_speed": 3.0066553328561203,
                    "wind_direction": 244.99999999999997,
                    "air_pressure": 1018.0,
                    "humidity": 60,
                    "visibility": 9.887879782072694,
                    "predictability": 77
                }
            ],
            "time": "2021-02-07T00:15:04.671620-08:00",
            "sun_rise": "2021-02-07T07:04:47.083497-08:00",
            "sun_set": "2021-02-07T17:37:57.743533-08:00",
            "timezone_name": "LMT",
            "parent": {
                "title": "California",
                "location_type": "Region / State / Province",
                "woeid": 2347563,
                "latt_long": "37.271881,-119.270233"
            },
            "sources": [
                {
                    "title": "BBC",
                    "slug": "bbc",
                    "url": "http://www.bbc.co.uk/weather/",
                    "crawl_rate": 360
                },
                {
                    "title": "Forecast.io",
                    "slug": "forecast-io",
                    "url": "http://forecast.io/",
                    "crawl_rate": 480
                },
                {
                    "title": "Met Office",
                    "slug": "met-office",
                    "url": "http://www.metoffice.gov.uk/",
                    "crawl_rate": 180
                },
                {
                    "title": "OpenWeatherMap",
                    "slug": "openweathermap",
                    "url": "http://openweathermap.org/",
                    "crawl_rate": 360
                },
                {
                    "title": "Weather Underground",
                    "slug": "wunderground",
                    "url": "https://www.wunderground.com/?apiref=fc30dc3cd224e19b",
                    "crawl_rate": 720
                },
                {
                    "title": "World Weather Online",
                    "slug": "world-weather-online",
                    "url": "http://www.worldweatheronline.com/",
                    "crawl_rate": 360
                }
            ],
            "title": "San Jose",
            "location_type": "City",
            "woeid": 2488042,
            "latt_long": "37.338581,-121.885567",
            "timezone": "America/Los_Angeles"
        }
    ]
}