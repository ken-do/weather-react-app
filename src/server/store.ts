/* eslint-disable global-require, @typescript-eslint/no-var-requires */
import { configureStore } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
import rootReducer from 'store/rootReducer'
import rootSaga from 'store/rootSaga'

const sagaMiddleware = createSagaMiddleware()

const store = configureStore({
    reducer: rootReducer,
    middleware: [sagaMiddleware],
})

sagaMiddleware.run(rootSaga)

export type AppDispatch = typeof store.dispatch

export default store
