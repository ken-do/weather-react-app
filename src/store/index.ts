/* eslint-disable global-require, @typescript-eslint/no-var-requires */
import { configureStore } from '@reduxjs/toolkit'
import logger from 'redux-logger'
import createSagaMiddleware from 'redux-saga'
import rootReducer from './rootReducer'

const sagaMiddleware = createSagaMiddleware()

const store = configureStore({
    reducer: rootReducer,
    middleware: [sagaMiddleware, logger] as const,
})

if (process.env.NODE_ENV === 'development' && module.hot) {
    module.hot.accept('./rootReducer', () => {
        const newRootReducer = require('./rootReducer').default
        store.replaceReducer(newRootReducer)
    })
}

export type AppDispatch = typeof store.dispatch

export default store
