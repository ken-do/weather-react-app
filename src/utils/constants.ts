// eslint-disable-next-line import/prefer-default-export
export const DEBOUNCE_WAIT_MS = 500
export const PATH = {
    home: '/',
    location: '/location',
    weather: '/weather',
}
export const MAX_SUGGESTIONS = 5
export const MAX_FORECASTS = 5
export const STATIC_IMG_ROOT = `${process.env.REACT_APP_API_ORIGIN}/static/img/weather/`
export const APP_NAME = 'Weather React App'
export const LOCATION_SEARCH_PLACEHOLDER = 'Enter a location...'
export const EVENT_KEYS = {
    enter: 'Enter',
}
