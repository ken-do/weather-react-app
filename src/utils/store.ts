import { useSelector, useDispatch, TypedUseSelectorHook } from 'react-redux'
import { AppDispatch } from 'store'
import { RootState } from 'store/rootReducer'

const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
const useAppDispatch = () => useDispatch<AppDispatch>()

export { useAppDispatch as useDispatch }
export { useAppSelector as useSelector }
