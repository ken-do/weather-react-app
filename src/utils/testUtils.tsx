import React, { PropsWithChildren, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import store from 'src/store'
import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux'

export const AllTheProviders = ({ children }: PropsWithChildren<unknown>) => (
    <Provider store={store}>
        <Router>{children}</Router>
    </Provider>
)

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'queries'>
) => {
    return render(ui, { wrapper: AllTheProviders, ...options })
}

export * from '@testing-library/react'

export { customRender as render }
