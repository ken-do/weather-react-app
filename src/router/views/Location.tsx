import React, { useEffect } from 'react'
import LocationList from 'features/location/LocationList'
import { useQuery } from 'utils/router'
import { useDispatch, useSelector } from 'utils/store'
import { getLocations } from 'features/location/locationsSlice'
import { EmptyStateMessage, LoadingIndicator } from 'components'

const Location = () => {
    const queryParams = useQuery()
    const query = queryParams.get('query')
    const dispatch = useDispatch()
    const locations = useSelector((state) => state.locations.entries)
    const isLoading = useSelector((state) => state.locations.isLoading)

    useEffect(() => {
        if (query) {
            dispatch(getLocations(query))
        }
    }, [dispatch, query])

    if (isLoading) {
        return <LoadingIndicator />
    }

    if (locations.length) {
        return <LocationList locations={locations} />
    }

    return (
        <EmptyStateMessage
            heading="No Results"
            subHeading="Try searching for another location with a different query."
        />
    )
}

export default Location
