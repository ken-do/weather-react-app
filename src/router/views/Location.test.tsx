import React from 'react'
import { render, waitFor } from 'src/utils/testUtils'
import store from 'src/store'
import {
    getLocationsSuccess,
    getLocationsFailure,
    resetLocations,
    getLocationsStart,
} from 'src/features/location/locationsSlice'
import { search as locations } from 'src/server/db.json'
import * as slice from 'src/features/location/locationsSlice'
import { PATH } from 'src/utils/constants'
import Location from './Location'

describe('Location view', () => {
    afterEach(() => {
        store.dispatch(resetLocations())
    })

    test('should render a loading indicator when location data is being fetched', () => {
        store.dispatch(getLocationsStart())
        const { queryByRole, asFragment } = render(<Location />)
        const loadingIndicator = queryByRole('status')
        expect(loadingIndicator).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    test('should show a list of locations after location data was', () => {
        store.dispatch(getLocationsSuccess(locations))
        const { queryAllByRole, asFragment } = render(<Location />)
        const locationList = queryAllByRole('article')
        expect(locationList).toHaveLength(locations.length)
        expect(asFragment()).toMatchSnapshot()
    })

    test('should show a message saying there is no results by default or if the request failed to return any data', () => {
        const { getByText, rerender, asFragment } = render(<Location />)
        const noResultsMessage = getByText(/no results/i)
        expect(noResultsMessage).toBeInTheDocument()
        store.dispatch(getLocationsSuccess([]))
        rerender(<Location />)
        const stillNoResultsMessage = getByText(/no results/i)
        expect(stillNoResultsMessage).toBeInTheDocument()
        store.dispatch(getLocationsFailure(new Error('Error fetching')))
        rerender(<Location />)
        const stillAgainNoResultsMessage = getByText(/no results/i)
        expect(stillAgainNoResultsMessage).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    test('should fetch locations when there is a "query" param', () => {
        const { rerender, asFragment } = render(<Location />)
        const url = `${PATH.location}?query=san`
        window.history.pushState({}, '', url)
        const spy = jest.spyOn(slice, 'getLocations')
        rerender(<Location />)
        waitFor(() => {
            expect(spy).toHaveBeenCalledWith()
        })
        spy.mockRestore()
        expect(asFragment()).toMatchSnapshot()
    })
})
