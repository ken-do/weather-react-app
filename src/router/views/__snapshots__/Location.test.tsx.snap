// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location view should fetch locations when there is a "query" param 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="mr-3 pr-3 align-top border-right inline-block align-content-center"
    >
      No Results
    </h1>
    <div
      class="inline-block align-middle"
    >
      <h2
        class="font-weight-normal lead"
        id="desc"
      >
        Try searching for another location with a different query.
      </h2>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Location view should render a loading indicator when location data is being fetched 1`] = `
<DocumentFragment>
  <div
    aria-label="Loading"
    role="status"
    style="text-align: center;"
  >
    <div
      class="m-1 spinner-grow text-primary"
    />
    <div
      class="m-1 spinner-grow text-secondary"
    />
    <div
      class="m-1 spinner-grow text-success"
    />
    <div
      class="m-1 spinner-grow text-danger"
    />
    <div
      class="m-1 spinner-grow text-warning"
    />
    <div
      class="m-1 spinner-grow text-info"
    />
    <div
      class="m-1 spinner-grow text-dark"
    />
  </div>
</DocumentFragment>
`;

exports[`Location view should show a list of locations after location data was 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/2487956"
      >
        San Francisco
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/2487889"
      >
        San Diego
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/2488042"
      >
        San Jose
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/2487796"
      >
        San Antonio
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/2488853"
      >
        Santa Cruz
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/349859"
      >
        Santiago
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/56558361"
      >
        Santorini
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/773964"
      >
        Santander
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/1132447"
      >
        Busan
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/773692"
      >
        Santa Cruz de Tenerife
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
    <article
      class="col-md-3"
    >
      <a
        class="link mr-2 mb-2 d-inline-block"
        href="/weather/2488867"
      >
        Santa Fe
      </a>
      <span
        class="badge badge-pill badge-success"
      >
        City
      </span>
    </article>
  </div>
</DocumentFragment>
`;

exports[`Location view should show a message saying there is no results by default or if the request failed to return any data 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="mr-3 pr-3 align-top border-right inline-block align-content-center"
    >
      No Results
    </h1>
    <div
      class="inline-block align-middle"
    >
      <h2
        class="font-weight-normal lead"
        id="desc"
      >
        Try searching for another location with a different query.
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
